### PR TITLE
Personal and transactional emails do not contain unsubscribe header

### PR DESCRIPTION
**Pull request by volha-pivavarchyk**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/136
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 

#### Description:

This PR removes an unsubscribe header from an individual and transactional emails. 
[topic in the forum](https://forum.mautic.org/t/personal-email-should-not-contain-an-unsubscribe-header-custom-email/18525/11#post_12)

https://forum.mautic.org/t/personal-email-should-not-contain-an-unsubscribe-header-custom-email/18525/11#post_12

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Send an individual email, an email via API `/emails/ID/contact/CONTACT_ID/send`, transactional email by form and by campaign.
3. Check the headers of the emails. There is no 'List-Unsubscribe'. 
4. Send a marketing email by form and by campaign.
5. Check the headers of the emails. There is 'ListUnsubscribe'.

### CODE REVIEW

**Comment by npracht** [comment]
Same feedback as https://github.com/mautic/mautic/pull/11786#issuecomment-1338941909. This is BC and for transactional at least, we cannot do it so simply. We need to offer a third type of campaign email.

**Comment by escopecz** [comment]
@npracht the BC break is because it's set as true by default? If it would be false by default then it wouldn't be a BC break right?

**Comment by npracht** [comment]
I agree @escopecz !

**Review by RCheesley — CHANGES_REQUESTED** [review]
Testing on Gitpod, when I send to an individual contact from their profile I get the following error in the logs, and it never sends:

```
[2022-12-12 15:26:00] mautic.CRITICAL: Uncaught PHP Exception Error: "Call to a member function getEmailType() on null" at /var/www/html/app/bundles/EmailBundle/Helper/MailHelper.php line 1510 {"exception":"[object] (Error(code: 0): Call to a member function getEmailType() on null at /var/www/html/app/bundles/EmailBundle/Helper/MailHelper.php:1510)
[stacktrace]
#0 /var/www/html/app/bundles/EmailBundle/Event/EmailSendEvent.php(316): Mautic\\EmailBundle\\Helper\\MailHelper->getCustomHeaders()
#1 /var/www/html/app/bundles/EmailBundle/EventListener/DateTimeTokenSubscriber.php(48): Mautic\\EmailBundle\\Event\\EmailSendEvent->getTextHeaders()
#2 /var/www/html/vendor/symfony/event-dispatcher/Debug/WrappedListener.php(126): Mautic\\EmailBundle\\EventListener\\DateTimeTokenSubscriber->onEmailGenerate(Object(Mautic\\EmailBundle\\Event\\EmailSendEvent), 'mautic.email_on...', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#3 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(251): Symfony\\Component\\EventDispatcher\\Debug\\WrappedListener->__invoke(Object(Mautic\\EmailBundle\\Event\\EmailSendEvent), 'mautic.email_on...', Object(Symfony\\Component\\HttpKernel\\Debug\\TraceableEventDispatcher))
#4 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(73): Symfony\\Component\\EventDispatcher\\EventDispatcher->callListeners(Array, 'mautic.email_on...', Object(Mautic\\EmailBundle\\Event\\EmailSendEvent))
#5 /var/www/html/vendor/symfony/event-dispatcher/Debug/TraceableEventDispatcher.php(168): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch(Object(Mautic\\EmailBundle\\Event\\EmailSendEvent), 'mautic.email_on...')
#6 /var/www/html/app/bundles/EmailBundle/Helper/MailHelper.php(1644): Symfony\\Component\\EventDispatcher\\Debug\\TraceableEventDispatcher->dispatch(Object(Mautic\\EmailBundle\\Event\\EmailSendEvent), 'mautic.email_on...')
#7 /var/www/html/app/bundles/EmailBundle/Helper/MailHelper.php(378): Mautic\\EmailBundle\\Helper\\MailHelper->dispatchSendEvent()
#8 /var/www/html/app/bundles/LeadBundle/Controller/LeadController.php(1409): Mautic\\EmailBundle\\Helper\\MailHelper->send(true, false, false)
#9 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(442): Mautic\\LeadBundle\\Controller\\LeadController->emailAction('55', '')
#10 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(159): Mautic\\CoreBundle\\Controller\\CommonController->executeAction('email', '55', 0, '')
#11 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw(Object(Symfony\\Component\\HttpFoundation\\Request), 1)
#12 /var/www/html/vendor/symfony/http-kernel/Kernel.php(201): Symfony\\Component\\HttpKernel\\HttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#13 /var/www/html/app/AppKernel.php(105): Symfony\\Component\\HttpKernel\\Kernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#14 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#15 /var/www/html/app/middlewares/HSTSMiddleware.php(35): Mautic\\Middleware\\CORSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#16 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\HSTSMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#17 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(52): Mautic\\Middleware\\CatchExceptionMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#18 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#19 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#20 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle(Object(Symfony\\Component\\HttpFoundation\\Request), 1, true)
#21 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle(Object(Symfony\\Component\\HttpFoundation\\Request))
#22 /var/www/html/index_dev.php(25): Stack\
un(Object(Stack\\StackedHttpKernel))
#23 {main}
"} {"hostname":"mautic-web","pid":3285}
```

~~Also, when testing the Transactional mail with a campaign the link unsubscribe is still present in the emails sent - see https://watch.screencastify.com/v/gPvz3mSPuzkYuATl37iP.~~ I did not realise that the default was to send with the list-unsubscribe to maintain BC - switched it over to No (new behaviour) and it works as expected - the list-unsubscribe header is removed from the transactional mail.

I believe that we should default the behaviour to what the industry standard dictates (transactional emails are not using the list unsubscribe header) with a lot of education before the release, and documentation about what it all means. This way we are following expected best practice with Mautic's default behaviour, and folks can always choose to change this globally by changing the setting.

**Comment by volha-pivavarchyk** [comment]
@RCheesley I fixed the error, and the issue with the marketing type of emails.

File: app/bundles/EmailBundle/Helper/MailHelper.php [hunk 1]
```diff
@@ -1506,6 +1518,13 @@ public function getCustomHeaders()
     {
         $headers = array_merge($this->headers, $this->getSystemHeaders());
 
+        // Personal and transactional emails do not contain unsubscribe header
+        $email                = $this->getEmail();
+        $addUnsubscribeHeader = $this->factory->getParameter('add_mass_mailing_header');
+        if (empty($email) || ('transactional' === $this->getEmailType() && !$addUnsubscribeHeader)) {
```
Comment (kuzmany) [inline app/bundles/EmailBundle/Helper/MailHelper.php]: If we know Email entity in the step, then we do not need store emailType. Do I miss something?

Comment (volha-pivavarchyk) [reply app/bundles/EmailBundle/Helper/MailHelper.php]: EmailType is not stored in the Email entity. It is taken from the campaign.

**Comment by jos0405** [comment]
I tested, and I see unsub header still in the email:
Created transactional email campaign step:
image

The email arrived has this header:
image

Marketing email also has the same header:
image

I also tested what happens if I remove the unsub from the bottom - the unsub header is there regardless :(

**Comment by jos0405** [comment]
Testing direct emails: that actually works perfectly!

image

**Review by jos0405 — CHANGES_REQUESTED** [review]
Hi,
Campaign action Send Email / Transactional email still has the unsub headers :(

**Comment by volha-pivavarchyk** [comment]
@jos0405 Could you please check settings (step 6). Whether or not the setting is disabled.

**Comment by jos0405** [comment]
Sorry my bad.
All good Approved

**Review by jos0405 — DISMISSED** [review]
Works like a charm

**Comment by escopecz** [comment]
Do we need the special config option? The headers should be removed from transactional emails by default in my opinion. And we can do such change as this will be released in major version. Just have to document it in the changelog

**Comment by RCheesley** [comment]
Can we also ensure that this works with sending via the API as I believe someone reported that it didn't, and I'm not sure if anyone has checked that! :)

**Comment by volha-pivavarchyk** [comment]
@RCheesley  Im so sorry, I really didn't check the API. I'll check it as soon as I can.

**Comment by volha-pivavarchyk** [comment]
@RCheesley I checked and fixed the issue. I removed the header for emails sending via the API `/emails/ID/contact/CONTACT_ID/send`.
To be sure. Are all emails sent via API `/emails/ID/contact/CONTACT_ID/send` 'transactional'?

**Comment by mabumusa1** [comment]
@volha-pivavarchyk I hope we can merge this PR today, I noticed there are still some opened conversations with Ruth and Kuzmany. are those resolved?

How can we push this PR to the finish line?

**Comment by nick-vanpraet** [comment]
Via the UI everything works, but via the API I'm still seeing a List-Unsubscribe header even after enabling the setting to not add it to Transactional emails. BUT, aren't all emails sent directly to a contact considered "personal/individual" emails and shouldn't get that header regardless of the new setting?

If I select the template of a transactional email via the UI when sending to contact, no unsub header (regardless of setting).
If I do the same via the API (I imagine that's the API equivalent?) I do get an unsub header.

**Comment by volha-pivavarchyk** [comment]
@nick-vanpraet Could you retest this PR, please. I've just fixed it and updated. Looks like I was fixing at the same time you were testing it :)

**Comment by volha-pivavarchyk** [comment]
@mabumusa1 I've removed a List-Unsubscribe header from emails sent via API.

**Comment by volha-pivavarchyk** [comment]
@jos0405 Could you please retest this PR. I've updated it.

**Comment by volha-pivavarchyk** [comment]
@escopecz I removed the setting and documented the changes. Could you please let me know if I need to detail the changes.

**Comment by nick-vanpraet** [comment]
> @nick-vanpraet Could you retest this PR, please. I've just fixed it and updated. Looks like I was fixing at the same time you were testing it :)

Tested API again, no more unsubscribe header!

**Comment by RCheesley** [comment]
I think we are RTC with this one - we do need to make sure that we have the right educational materials for this change when we release. If anyone would like to help with this by contributing content please let us know.

We have https://github.com/mautic/user-documentation/pull/136 for the end-user docs (has pending feedback to resolve) but I think we need to have also KB articles and blog articles to support. cc @tobsowo @deborahsalves @fakela for information - the changes are definitely going to be in 5.0.

**Comment by mabumusa1** [comment]
> I think we are RTC with this one - we do need to make sure that we have the right educational materials for this change when we release. If anyone would like to help with this by contributing content please let us know.
> 
> We have [mautic/user-documentation#136](https://github.com/mautic/user-documentation/pull/136) for the end-user docs (has pending feedback to resolve) but I think we need to have also KB articles and blog articles to support. cc @tobsowo @deborahsalves @fakela for information - the changes are definitely going to be in 5.0.

There is a good document here https://mautic.slack.com/archives/CQMKV0RU1/p1676049406854669 about this PR

**Comment by RCheesley** [comment]
Good to merge once the tests complete.

**Pull request merged**